File: Tags/src/main/java/com/aliucord/plugins/Tags.java
/*
 * Copyright (c) 2021 Juby210
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.plugins;

import android.content.Context;

import androidx.annotation.NonNull;

import com.aliucord.Utils;
import com.aliucord.api.CommandsAPI;
import com.aliucord.entities.Plugin;
import com.aliucord.plugins.tags.*;
import com.discord.api.commands.ApplicationCommandType;
import com.discord.api.commands.CommandChoice;
import com.discord.models.commands.ApplicationCommandOption;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Tags extends Plugin {
    @NonNull
    @Override
    public Manifest getManifest() {
        Manifest manifest = new Manifest();
        manifest.authors = new Manifest.Author[]{ new Manifest.Author("Juby210", 324622488644616195L) };
        manifest.description = "Allows you to send custom tags.";
        manifest.version = "1.0.5";
        manifest.updateUrl = "https://raw.githubusercontent.com/Juby210/Aliucord-plugins/builds/updater.json";
        return manifest;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void start(Context context) {
        existingTags = new ArrayList<>();
        subcommands = new ArrayList<>();

        ApplicationCommandOption tagName = new ApplicationCommandOption(ApplicationCommandType.STRING, "name", "Tag name",
                null, true, true, null, null);
        ApplicationCommandOption existingTag = new ApplicationCommandOption(ApplicationCommandType.STRING, "name", "Tag name",
                null, true, true, existingTags, null);

        subcommands.add(new ApplicationCommandOption(
                ApplicationCommandType.SUBCOMMAND,
                "add",
                "Create a tag",
                null,
                false,
                false,
                null,
                Arrays.asList(tagName, CommandsAPI.requiredMessageOption)
        ));
        subcommands.add(new ApplicationCommandOption(
                ApplicationCommandType.SUBCOMMAND,
                "delete",
                "Delete a tag",
                null,
                false,
                false,
                null,
                Collections.singletonList(existingTag)
        ));
        subcommands.add(new ApplicationCommandOption(
                ApplicationCommandType.SUBCOMMAND,
                "list",
                "Shows list with tag names",
                null,
                false,
                false,
                null,
                Collections.emptyList()
        ));
        subcommands.add(new ApplicationCommandOption(
                ApplicationCommandType.SUBCOMMAND,
                "rename",
                "Rename a tag",
                null,
                false,
                false,
                null,
                Arrays.asList(
                        existingTag,
                        new ApplicationCommandOption(ApplicationCommandType.STRING, "newName", "New tag name",
                                null, true, true, null, null)
                )
        ));
        subcommands.add(new ApplicationCommandOption(
                ApplicationCommandType.SUBCOMMAND,
                "update",
                "Update a tag",
                null,
                false,
                false,
                null,
                Arrays.asList(existingTag, CommandsAPI.requiredMessageOption)
        ));

        HashMap<String, String> _tags = sets.getObject("tags", null, tagsType);
        if (_tags != null) for (Map.Entry<String, String> tag : _tags.entrySet()) registerTag(tag.getKey(), tag.getValue());

        commands.registerCommand(
                "tag",
                "Send and manage tags",
                subcommands,
                args -> {
                    if (args.containsKey("add")) return AddCommand.execute((Map<String, ?>) args.get("add"), sets, this);
                    if (args.containsKey("delete")) return DeleteCommand.execute((Map<String, ?>) args.get("delete"), sets, this);
                    if (args.containsKey("list")) return ListCommand.execute(sets);
                    if (args.containsKey("rename")) return RenameCommand.execute((Map<String, ?>) args.get("rename"), sets, this);
                    if (args.containsKey("update")) return UpdateCommand.execute((Map<String, ?>) args.get("update"), sets, this);

                    HashMap<String, String> tags = sets.getObject("tags", null, tagsType);
                    if (tags != null) for (Map.Entry<String, String> tag : tags.entrySet()) {
                        if (args.containsKey(tag.getKey()))
                            return new CommandsAPI.CommandResult(runTag(tag.getValue(), (Map<String, ?>) args.get(tag.getKey())));
                    }

                    return new CommandsAPI.CommandResult();
                }
        );
    }

    @Override
    public void stop(Context context) {
        commands.unregisterAll();
    }

    public static final Type tagsType = TypeToken.getParameterized(HashMap.class, String.class, String.class).getType();

    public final Pattern argPattern = Pattern.compile("\\[(\\w+)]");
    public List<CommandChoice> existingTags;
    public List<ApplicationCommandOption> subcommands;

    public void registerTag(String name, String message) {
        ArrayList<ApplicationCommandOption> args = new ArrayList<>();
        Matcher matcher = argPattern.matcher(message);
        while (matcher.find()) {
            String argName = matcher.group(1);
            args.add(new ApplicationCommandOption(
                    ApplicationCommandType.STRING,
                    argName,
                    argName + " content",
                    null,
                    false,
                    false,
                    null,
                    null
            ));
        }

        existingTags.add(Utils.createCommandChoice(name, name));
        subcommands.add(new ApplicationCommandOption(
                ApplicationCommandType.SUBCOMMAND,
                name,
                "Tag: " + (message.length() > 150 ? message.substring(0, 150) + "…" : message),
                null,
                false,
                false,
                null,
                args
        ));
    }

    public String runTag(String value, Map<String, ?> args) {
        Matcher matcher = argPattern.matcher(value);
        while (matcher.find()) {
            String v = (String) args.get(matcher.group(1));
            value = value.replace(matcher.group(), v == null ? "" : v);
        }
        return value;
    }
}

File: Tags/src/main/java/io/github/juby210/acplugins/Tags.kt
/*
 * Copyright (c) 2021 Juby210
 * Licensed under the Open Software License version 3.0
 */

package io.github.juby210.acplugins

import android.content.Context
import com.aliucord.Utils.createCommandChoice
import com.aliucord.Utils.createCommandOption
import com.aliucord.annotations.AliucordPlugin
import com.aliucord.api.CommandsAPI
import com.aliucord.entities.Plugin
import com.discord.api.commands.ApplicationCommandType
import com.discord.api.commands.CommandChoice
import com.discord.models.commands.ApplicationCommandOption
import com.google.gson.reflect.TypeToken
import io.github.juby210.acplugins.tags.*
import java.lang.reflect.Type
import java.util.*
import java.util.regex.Pattern
import kotlin.collections.ArrayList

@AliucordPlugin
class Tags : Plugin() {
    override fun start(context: Context) {
        val tagName = createCommandOption(
            name = "name",
            description = "Tag name",
            required = true,
            default = true
        )
        val existingTag = createCommandOption(
            name = "name",
            description = "Tag name",
            required = true,
            default = true,
            choices = existingTags
        )

        subcommands.add(createCommandOption(
            type = ApplicationCommandType.SUBCOMMAND,
            name = "add",
            description = "Create a tag",
            subCommandOptions = listOf(tagName, CommandsAPI.requiredMessageOption)
        ))
        subcommands.add(createCommandOption(
            type = ApplicationCommandType.SUBCOMMAND,
            name = "delete",
            description = "Delete a tag",
            subCommandOptions = listOf(existingTag)
        ))
        subcommands.add(createCommandOption(
            type = ApplicationCommandType.SUBCOMMAND,
            name = "list",
            description = "Shows list with tag names"
        ))
        subcommands.add(createCommandOption(
            type = ApplicationCommandType.SUBCOMMAND,
            name = "rename",
            description = "Rename a tag",
            subCommandOptions = listOf(
                existingTag,
                createCommandOption(
                    name = "newName",
                    description = "New tag name",
                    required = true,
                    default = true
                )
            )
        ))
        subcommands.add(createCommandOption(
            type = ApplicationCommandType.SUBCOMMAND,
            name = "update",
            description = "Update a tag",
            subCommandOptions = listOf(existingTag, CommandsAPI.requiredMessageOption)
        ))

        val tags = settings.getObject<HashMap<String, String>?>("tags", null, tagsType)
        if (tags != null) for ((key, value) in tags) registerTag(key, value)
        commands.registerCommand(
            "tag",
            "Send and manage tags",
            subcommands
        ) { ctx ->
            return@registerCommand if (ctx.containsArg("add")) AddCommand.execute(ctx.getSubCommandArgs("add"), settings, this)
            else if (ctx.containsArg("delete")) DeleteCommand.execute(ctx.getSubCommandArgs("delete"), settings, this)
            else if (ctx.containsArg("list")) ListCommand.execute(settings)
            else if (ctx.containsArg("rename")) RenameCommand.execute(ctx.getSubCommandArgs("rename"), settings, this)
            else if (ctx.containsArg("update")) UpdateCommand.execute(ctx.getSubCommandArgs("update"), settings, this)
            else {
                settings.getObject<HashMap<String, String>?>("tags", null, tagsType)?.run {
                    for ((key, value) in this)
                        if (ctx.containsArg(key)) return@registerCommand CommandsAPI.CommandResult(runTag(value, ctx.getSubCommandArgs(key)))
                }
                CommandsAPI.CommandResult()
            }
        }
    }

    override fun stop(context: Context) {
        commands.unregisterAll()
        existingTags.clear()
        subcommands.clear()
    }

    private val argPattern = Pattern.compile("\\[(\\w+)]")
    @JvmField
    val existingTags = ArrayList<CommandChoice>()
    @JvmField
    val subcommands = ArrayList<ApplicationCommandOption>()

    fun registerTag(name: String, message: String) {
        val args = ArrayList<ApplicationCommandOption>()
        val matcher = argPattern.matcher(message)
        while (matcher.find()) {
            val argName = matcher.group(1)
            args.add(createCommandOption(
                name = argName!!,
                description = "$argName content"
            ))
        }
        existingTags.add(createCommandChoice(name, name))
        subcommands.add(createCommandOption(
            type = ApplicationCommandType.SUBCOMMAND,
            name = name,
            description = "Tag: " + if (message.length > 150) message.substring(0, 150) + "…" else message,
            subCommandOptions = args
        ))
    }

    private fun runTag(value: String, args: Map<String?, *>?): String {
        var ret = value
        val matcher = argPattern.matcher(ret)
        while (matcher.find()) {
            val v = args!![matcher.group(1)] as String?
            ret = ret.replace(matcher.group(), v ?: "")
        }
        return ret
    }

    companion object {
        @JvmField
        val tagsType: Type = TypeToken.getParameterized(HashMap::class.java, String::class.java, String::class.java).getType()
    }
}
